### PR TITLE
Replace toasts with improved custom optimistic UX

### LINKED ISSUE #849: External pending payment UX
**tldr Toasts are perfectly okay for now - and maybe the best thing to do while we're experimenting - but we'll want something better once we nail the UX otherwise**

Some good discussion here: https://stacker.news/items/423928/r/k00b?commentId=428876

In addition to #848, we'll want something more refined than toasts as time goes on to facilitate cancels/retries. Basically these actions will be fundamental parts of the non-custodial UX so we shouldn't offload them to generic toast messages in the longrun.

### PR DESCRIPTION
**Pull request by ekzyis**
## Description

Close #849 Related to #848

As mentioned in #1071, this replaces `useInvoiceable` with `usePayment` which has a simpler API.

This PR also contains changes to improve the external payment UX by replacing generic toasts with custom UX code per action.

TODO:
- [x] #1071
- [x] optimistic zap UX
- [x] optimistic custom zap UX (custom zap = zapping via long press)
- [x] optimistic reply UX
- [x] ~optimistic post UX~ _canceled because this is most likely better implemented using server-side code so let's first get user feedback on client-side optimistic update code before delving deeper into UX patterns that include server-side code_
  - [x] ~bounty form~
  - [x] ~discussion form~
  - [x] ~link form~
  - [x] ~poll form~
  - [x] ~job form~
- [x] optimistic poll votes UX
- [x] optimistic bounty payments UX
- [x] ~territory admin optimistic UX~
  - [x] ~create~
  - [x] ~update~
  - [x] ~unarchive~
- [x] (zap) undos
- [x] remove unnecessary payment context (see [this comment](https://github.com/stackernews/stacker.news/pull/1161#discussion_r1593327834))
- [x] fix `meAnonSats` not immediately updated

**Error handling**

- [x] inform users about errors in notifications to let them retry
- [x] ~remove fallback to QR code on attached wallet payment failure since it interrupts UX (rely on out-of-band error handling like notifications instead) (?)~ _let's see in production first_
- [x] fix `Cannot read properties of null (reading 'meSats')` when reloading page on /notifications and trying to zap error notification because cache is not populated
- [x] fix client notifications don't use cache but static item data. this means optimistic updates don't work for client notifications.

## Screenshots

Since toasts for success _and_ error were removed for better UX, we now use ~dismissible~ error notifications that are only stored locally. So if you zap something and it fails, it looks like this:

localhost_3000_(iPhone SE)

## Additional Context

See https://github.com/stackernews/stacker.news/pull/1071#issuecomment-2094562629 for additional context regarding move from `useInvoiceable` to `usePayment`. Most notably, `JITInvoice` is no longer with us. I removed it because it made the code more complicated and the retry/cancel UX was bad and lead regularly to confusion ("sats received" vs "error"). The HODL invoice is now immediately canceled if the action fails. With the new custom optimistic UX, retries via literally retrying (pressing `reply`, zapping again etc.) are imo simple enough.

## Checklist

**Are your changes backwards compatible? Please answer below:**

**Did you QA this? Could we deploy this straight to production? Please answer below:**

**For frontend changes: Tested on mobile? Please answer below:**

**Did you introduce any new environment variables? If so, call them out explicitly here:**

### CODE REVIEW

File: components/payment.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  return (
+    <PaymentContext.Provider value={value}>
+      {children}
+    </PaymentContext.Provider>
+  )
+}
+
+export const usePayment = () => {
+  return useContext(PaymentContext)
+}
```
Comment (ekzyis) [inline components/payment.js:211]: Mhh, I don't really need to wrap the functions in a context since I do nothing with it. I thought I might need to be context aware but even better that I actually don't. I always return the same two functions. Guess I can remove it and move all code from `PaymentProvider` to `usePayment`.

**Comment by ekzyis** [comment]
Random better idea for zap undos: instead of this ugly orange toast with "undo" written inside, use the pending pulse animation I had before and use that as undo.

So while the bolt is pulsing, you can click it again to undo.

_But if you're too slow, you'll zap your default, lol._

File: components/notifications.js [hunk 2]
```diff
@@ -566,3 +597,153 @@ function CommentsFlatSkeleton () {
...
+
+const NotificationContext = createContext()
+
+function loadNotifications (storageKey) {
+  const stored = window.localStorage.getItem(storageKey)
+  if (!stored) return []
+  const filtered = JSON.parse(stored).filter(({ sortTime }) => {
+    // only keep notifications younger than 24 hours
+    return new Date(sortTime) >= datePivot(new Date(), { hours: -24 })
+  })
```
Comment (ekzyis) [inline components/notifications.js:609]: > **Only keep notifications younger than 24 hours**
>
> This helps with not bloating /notifications and local storage.
>
> It also is a workaround for having to come up with some complicated pagination algorithm which considers how many / which client notifications should be included on each page.

-- 5ff6210e4ffa6d20b64918b822a9ad81f0ae0c10

_I think this is smart lazy stuff_

File: components/toast.module.css [hunk 3]
```diff
@@ -1,5 +1,6 @@
 .toastContainer {
   transform: translate3d(0, 0, 0);
+  margin-bottom: 53px; /* padding for bottom nav bar on mobile */
```
Comment (ekzyis) [inline components/toast.module.css:3]: unrelated fix for toasts shadowing bottom nav bar

File: components/notifications.js (hunk 2 above)
Comment (huumn) [reply components/notifications.js:609]: This is a reasonable shim for sure

**Review by huumn — COMMENTED** [review]
This looks really thoughtfully done on first pass. I'll do a deep review and QA the first chance I get tomorrow.

**Comment by huumn** [comment]
Somehow I end up increasing my balance with this approach. Aren't those invoices not supposed to settle?

Also, one of my zaps ended up showing up as a flag (on the sn wallet plans post). No idea what that's about.

**Review by huumn — CHANGES_REQUESTED** [review]
Failing QA

File: components/form.js [hunk 4]
```diff
@@ -820,37 +833,39 @@ export function Form ({
...
+          throw new SessionRequiredError()
+        }
+        beforeSubmitReturn = beforeSubmit?.({ amount, ...values })
+        let hash, hmac
+        if (invoiceable) {
+          revert = await optimisticUpdate?.({ amount, ...values }, ...args);
+          [{ hash, hmac }, cancel] = await payment.request(amount)
         }
-        await onSubmit(values, ...args)
+        await onSubmit({ hash, hmac, amount, ...values }, ...args)
```
[the remark this answers is not in the document]
Comment (ekzyis) [reply components/form.js:849]: I'm only adding hash, hmac, amount to values which was already an object. I don't think this changes the signature.

You might missed the curly braces around the comma separated values?

[the remark this answers is not in the document]
Comment (huumn) [reply components/form.js:849]: Oh right, values was already an object. What's in args?

**Comment by ekzyis** [comment]
> Somehow I end up increasing my balance with this approach. Aren't those invoices not supposed to settle?

You're right, they should either settle because they were consumed by an action or canceled. So in both cases, the balance shouldn't increase.

> Also, one of my zaps ended up showing up as a 0-sat flag (on the sn wallet plans post). No idea what that's about, but I suspect there's a bug somewhere that caused that.

I had this error before. It happens when a payment fails and the code that reverts the optimistic update ends up writing a negative value instead of the value before the update was done.

> Did you test all sending side wallets after these changes? It might help discover more unhappy paths.

No, I assumed that all sending wallets would behave the same since I only relied on `sendPayment` working afaict. But I guess there are some details I missed, most likely related to error handling.

File: components/form.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ekzyis) [reply components/form.js:849]: formik helper stuff like `resetForm`. They call it "formikBag"  in the [documentation](https://formik.org/docs/api/formik#onsubmit-values-values-formikbag-formikbag--void--promiseany). Maybe we should use that name, too. Since we no longer wrap submit handlers for our payment flow, I think `args` is really the same as `formikBag`.

**Comment by huumn** [comment]
> No, I assumed that all sending wallets would behave the same since I only relied on sendPayment working afaict. But I guess there are some details I missed, most likely related to error handling.

Kind of meta, but because all the wallets are all in a `beta` state, we should probably QA all of them when we make changes to related code. It'll also help us familiarize ourselves with the UX of each wallet over and over again, so that all of them work equally well. So it's both QA and UX investigation.

File: components/form.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (huumn) [reply components/form.js:849]: Oh I see. I remember using `resetForm` at one point.

Someone at Square told me they don't allow engineers to declare array indices as `i`. Kind of intense lol

**Comment by ekzyis** [comment]
> I need to get NWC working

#1151 should have you covered. You only need to install [`nostr-wallet-connect-lnd`](https://github.com/benthecarman/nostr-wallet-connect-lnd). It uses wss://relay.damus.io by default which has rate limits which can be annoying. But so far, it wasn't annoying enough to find a solution.

> add a LNBits container.

Yep, also thought about this. I'll do this today. I think this is more in my ball park anyway since I also use LNbits to manage my lightning credentials for hnbot and oracle + I need this to test this PR.

**Comment by ekzyis** [comment]
Mhh, I fixed a race condition in 3ddd6b4 but there is still one: when we zap too fast, pending sats are counted again for each zap since `cache.modify` triggers `readField` which then adds pending sats that are already included in existing sats _again_ to the existing sats.

For some reason I completely missed this before, I must have been too cautious in testing my code or NWC payments were too slow so I didn't notice this. I tried to fix this by accounting for existing pending sats when calling `cache.modify` but didn't work for some reason. So I guess there is more stuff broken in my assumptions.

This also isn't the only issue I found: We don't update `me.privates.sats` so there is also a bug where following zaps reuse the existing balance and thus think they don't need an invoice but then fail since they actually do.

I'm starting to believe we should straight go for storing pending states in the server and skip this release of client-side optimistic updates. On the server, we can use database transactions to make sure we don't run into race conditions like this. Feels bad that it took me so long to realize that doing this on the client is way too messy and buggy but better pivoting late than releasing this in its current state. Going to sleep over this but I think this is the right call. Doing all of this stuff on the server should be much simpler, thus cleaner and is what we wanted to do anyway.

**Comment by ekzyis** [comment]
Superseded by #1184
